### PR TITLE
Set up magnetic field in GIZMO frontend

### PR DESCRIPTION
**Pull request by qobilidop**
## PR Summary

A function to set up magnetic field aliases for particle dataset is implemented. And it is used to set up the magnetic field in GIZMO frontend.

## PR Checklist

- [x] Code passes flake8 checker
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds tests for new features

### CODE REVIEW

File: yt/fields/magnetic_field.py [hunk 1]
```diff
@@ -194,11 +194,61 @@ def setup_magnetic_field_aliases(registry, ds_ftype, ds_fields, ftype="gas"):
...
+        def _mag_field(field, data):
+            return convert(data[fd])
+        return _mag_field
+    for ax, fd in zip(registry.ds.coordinates.axis_order, ds_fields):
+        registry.add_field((ftype,"magnetic_field_%s" % ax),
+                           sampling_type="cell", 
+                           function=mag_field(fd),
+                           units=units)
+
+def setup_particle_magnetic_field_aliases(registry, ds_ptype, ds_field):
```
Comment (qobilidop) [inline yt/fields/magnetic_field.py]: I think for particle datasets we always want the resulting particle type to be the same as the particle type to be aliased. So the last argument is removed as compared with `setup_magnetic_field_aliases`. I'd like to know others' opinions on this.

File: yt/fields/magnetic_field.py [hunk 2]
```diff
@@ -210,12 +260,5 @@ def setup_magnetic_field_aliases(registry, ds_ftype, ds_fields, ftype="gas"):
...
         convert = lambda x: x.to_equivalent(to_units, equiv)
-    def mag_field(fd):
-        def _mag_field(field, data):
-            return convert(data[fd])
-        return _mag_field
-    for ax, fd in zip(registry.ds.coordinates.axis_order, ds_fields):
-        registry.add_field((ftype,"magnetic_field_%s" % ax),
-                           sampling_type="cell", 
-                           function=mag_field(fd),
-                           units=unit_system[to_units.dimensions])
```
Comment (qobilidop) [inline yt/fields/magnetic_field.py]: I don't quite understand the reason to use `unit_system[to_units.dimensions]` instead of `to_units` here. Could someone explain?

**Comment by qobilidop** [comment]
@tkc004 you might want to have a look.

**Comment by jzuhone** [comment]
Do we have a GIZMO dataset with magnetic fields we can test this on? I also think that since you're touching the magnetic field setup maybe we should beef up the tests here: https://github.com/yt-project/yt/blob/master/yt/fields/tests/test_magnetic_fields.py

Maybe add a third dataset after the two there and make `unit_system="code"`.

File: yt/fields/magnetic_field.py (hunk 1 above)
Comment (jzuhone) [reply yt/fields/magnetic_field.py]: Is there not a way we can make `setup_magnetic_field_aliases` do this job also by merging the functionality of these two functions into one and just give it appropriate arguments?

Comment (qobilidop) [reply yt/fields/magnetic_field.py]: Yes, I think it's doable.

File: yt/fields/magnetic_field.py (hunk 2 above)
Comment (qobilidop) [reply yt/fields/magnetic_field.py]: Actually, I have kept this in the refactored code. That's why the tests pass.

File: yt/fields/magnetic_field.py [hunk 3]
```diff
@@ -170,35 +170,56 @@ def _mach_alfven(field,data):
 
 def setup_magnetic_field_aliases(registry, ds_ftype, ds_fields, ftype="gas"):
```
Comment (qobilidop) [inline yt/fields/magnetic_field.py]: @jzuhone I have merged the treatments for grid and particle dataset into a single function as you suggested. Let me know if you have further suggestions.

**Comment by ngoldbaum** [comment]
Agreed that I'd like to see a test with a public test dataset before this gets merged. Otherwise LGTM :)

I'm glad this was relatively painless :)

File: yt/fields/tests/test_magnetic_fields.py [hunk 4]
```diff
@@ -19,19 +23,23 @@ def test_magnetic_fields():
...
     dd2 = ds2.all_data()
+    dd3 = ds3.all_data()
 
     assert ds1.fields.gas.magnetic_field_strength.units == "gauss"
     assert ds1.fields.gas.magnetic_field_poloidal.units == "gauss"
     assert ds1.fields.gas.magnetic_field_toroidal.units == "gauss"
     assert ds2.fields.gas.magnetic_field_strength.units == "T"
     assert ds2.fields.gas.magnetic_field_poloidal.units == "T"
     assert ds2.fields.gas.magnetic_field_toroidal.units == "T"
+    assert ds3.fields.PartType0.particle_magnetic_field.units == "code_magnetic"
```
Comment (qobilidop) [inline yt/fields/tests/test_magnetic_fields.py]: I'm not sure if this test is enough. The tests below cannot be performed on ds3 because I haven't added all the derived magnetic fields to the Gizmo frontend.

Comment (ngoldbaum) [reply yt/fields/tests/test_magnetic_fields.py]: I'd test to make sure you can access these fields and that e.g. the vector fields and component fields, as well as the aliases to internal yt field names are the correct shapes and have the correct units.

With the demeshening you'd be able to use the existing definitions for derived magnetic fields. Unfortunately I think to get those working here you'll need to define particle-specific versions of them.

Also I'd prefer it if this test were in the Gizmo frontend, even if it duplicates some boilerplate from the existing test you've modified here. The way you've modified this existing test means that the test will only run if the GIZMO MHD dataset is present in the dataset search path. That means that this test will no longer run on travis, for example.

Comment (qobilidop) [reply yt/fields/tests/test_magnetic_fields.py]: Thanks, I'll change them as you suggested.

**Comment by ngoldbaum** [comment]
It may be worth adding a note to the docs on loading Gizmo data that magnetic fields are supported via the `particle_magnetic_field` field.

**Comment by qobilidop** [comment]
@yt-fido test this please

**Comment by qobilidop** [comment]
@yt-fido test this please

**Comment by ngoldbaum** [comment]
@chummels can I get you to take a look at this PR?

File: yt/fields/field_detector.py [hunk 5]
```diff
@@ -126,9 +126,11 @@ def __missing__(self, item):
         elif finfo is not None and finfo.particle_type:
             if "particle_position" in (item, item[1]) or \
                "particle_velocity" in (item, item[1]) or \
+               "particle_magnetic_field" in (item, item[1]) or \
                "Velocity" in (item, item[1]) or \
                "Velocities" in (item, item[1]) or \
-               "Coordinates" in (item, item[1]):
+               "Coordinates" in (item, item[1]) or \
+               "MagneticField" in ((item, item[1])):
```
Comment (chummels) [inline yt/fields/field_detector.py]: Do you need to have the extra set of ()?

File: yt/fields/magnetic_field.py [hunk 6]
```diff
@@ -170,35 +170,56 @@ def _mach_alfven(field,data):
...
-    if ds_fields[0] not in registry:
+    if isinstance(ds_fields, list):
+        # If ds_fields is a list, we assume a grid dataset
+        sampling_type = "cell"
+        ds_fields = [(ds_ftype, fd) for fd in ds_fields]
+        ds_field = ds_fields[0]
+    else:
+        # Otherwise, we assume a particle dataset
+        sampling_type = "particle"
+        ds_field = ds_fields
```
Comment (chummels) [inline yt/fields/magnetic_field.py]: I'm not sure you need this line, as it's getting clobbered in the next line.

File: yt/fields/tests/test_magnetic_fields.py [hunk 7]
```diff
@@ -19,19 +19,26 @@ def test_magnetic_fields():
 
     ds1 = load_uniform_grid(data1, ddims, unit_system="cgs")
     ds2 = load_uniform_grid(data2, ddims, unit_system="mks")
+    # By default, code units are cgs units
```
Comment (chummels) [inline yt/fields/tests/test_magnetic_fields.py]: I think code units are only cgs units for gizmo, but given that this is only implemented for gizmo, I guess that is OK?

File: yt/fields/field_detector.py (hunk 5 above)
Comment (qobilidop) [reply yt/fields/field_detector.py]: That's a typo. Nice catch!

File: yt/fields/magnetic_field.py (hunk 6 above)
Comment (qobilidop) [reply yt/fields/magnetic_field.py]: I agree it would be cleaner that way.

File: yt/fields/tests/test_magnetic_fields.py (hunk 7 above)
Comment (qobilidop) [reply yt/fields/tests/test_magnetic_fields.py]: Actually, I'm not very familiar with how the code units are determined in general. I checked that for this test dataset, it's actually cgs. I'll make the comments more specific.
